lib/ex_rlp/decode.ex

```diff
@@ -3,83 +3,117 @@ defmodule ExRLP.Decode do
+    |> decode_item
   end
 
-  @spec unencode(binary() | String.t(), atom()) :: binary()
+  @spec unencode(binary(), atom()) :: binary()
   defp unencode(value, :binary), do: value
-  defp unencode(value, :hex), do: Base.decode16!(value, case: :lower)
+  defp unencode(value, :hex), do: decode_hex(value)
 
   @spec decode_item(binary(), ExRLP.t()) :: ExRLP.t()
-  defp decode_item(rlp_binary, result \\ [])
+  defp decode_item(rlp_binary, result \\ nil)
 
-  # When we don't have any RLP left, return the accumulator
-  defp decode_item(<<>>, result) do
+  defp decode_item("", result) when is_list(result) do
     Enum.reverse(result)
   end
 
-  # Decodes the head represents an item to be added directly
-  # to the result.
+  defp decode_item("", result) do
+    result
+  end
+
   defp decode_item(<<(<<prefix>>), tail::binary>>, result) when prefix < 128 do
-    decode_item(tail, [<<prefix>> | result])
+    new_item = <<prefix>>
+
+    new_result = add_new_item(result, new_item)
+
+    decode_item(tail, new_result)
   end
 
-  # Decodes medium length-binary?
   defp decode_item(<<(<<prefix>>), tail::binary>>, result) when prefix <= 183 do
     {new_item, new_tail} = decode_medium_binary(prefix, tail, 128)
 
-    decode_item(new_tail, [new_item | result])
+    new_result = add_new_item(result, new_item)
+
+    decode_item(new_tail, new_result)
   end
 
-  # Decodes long length-binary?
   defp decode_item(<<(<<be_size_prefix>>), tail::binary>>, result) when be_size_prefix < 192 do
     {new_item, new_tail} = decode_long_binary(be_size_prefix, tail, 183)
 
-    decode_item(new_tail, [new_item | result])
+    new_result = add_new_item(result, new_item)
+
+    decode_item(new_tail, new_result)
+  end
+
+  defp decode_item(<<(<<be_size_prefix>>), tail::binary>>, result) when be_size_prefix == 192 do
+    new_item = []
+
+    new_result = add_new_item(result, new_item)
+
+    decode_item(tail, new_result)
+  end
+
+  defp decode_item(<<(<<prefix>>), tail::binary>>, result) when prefix <= 247 do
+    {list, new_tail} = decode_medium_binary(prefix, tail, 192)
+
+    new_result = add_decoded_list(result, list)
+
+    decode_item(new_tail, new_result)
+  end
+
+  defp decode_item(<<(<<be_size_prefix>>), tail::binary>>, result) do
+    {list, new_tail} = decode_long_binary(be_size_prefix, tail, 247)
+
+    new_result = result |> add_decoded_list(list)
+
+    decode_item(new_tail, new_result)
   end
 
-  # Decodes an empty list
-  defp decode_item(<<(<<be_size_prefix>>), tail::binary>>, result)
-       when be_size_prefix == 192 do
-    decode_item(tail, [[] | result])
+  @spec add_new_item(ExRLP.t(), ExRLP.t()) :: ExRLP.t()
+  def add_new_item(nil, new_item) do
+    new_item
   end
 
-  defp decode_item(<<(<<prefix>>), tail::binary>>, result) do
-    {list_bin, new_tail} =
-      if prefix <= 247 do
-        decode_medium_binary(prefix, tail, 192)
-      else
-        decode_long_binary(prefix, tail, 247)
-      end
+  def add_new_item(result, new_item) do
```

Comment:
I'm not sure if this function is more clear than simply inline the list operation. Mostly, it obscures (in the code above) that this function is just a simple operation and nothing more complicated.

Reply:
yes, I agree. But we may have `nil` in the result variable.
I can change it to
```
if is_nil(result) do
  new_item
else
  [new_item | result]
end
```
in every place where this function is used